### ISSUE TITLE
Use firefox sync

### ISSUE THREAD

**Issue by aprokop** [kintesh/containerise#30]
Would it be possible to use `browser.storage.sync` instead of `browser.storage.local`?

**Comment by ethhics** [comment]
On first inspection I'd say it looks possible: the API looks similar, so you could simply change the two instances of `local` in the whole repo with `sync`. There's definitely appeal to having it work with FF sync, too.

**Comment by szalkerous** [comment]
PLEASE add in Sync Storage API support. I realize it's limited to 100KB per extension, but I'd be shocked if this one exceeds that even in heavy-use environments. Having to export and use hacky workarounds to sync the lists between installs is really rough and unnecessary. I noticed this has been open for almost 8 months... this is a make-it-or-break-it feature for me to continue using Containerise.

See https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/storage/sync for API usage.

**Comment by k4n30** [comment]
I'm all for using the firefox sync function, but it's not supported on all browsers, so it's going to require more effort than just replacing `local` with `sync`. It would also require some logic to handle unsupported browsers

**Comment by mcepl** [comment]
> I'm all for using the firefox sync function, but it's not supported on all browsers,

What? We are talking about extension for Firefox here. Which version of Firefox doesn't support Firefox Sync?

**Comment by k4n30** [comment]
Sorry I was referring to the sync api that is being proposed. I incorrectly assumed this extension worked on other browsers too, some of which the api works with and some it doesn't. I stand corrected

**Comment by mcepl** [comment]
Yeah, the repository is called “Firefox extension to automatically open websites in a container”.

**Comment by da2x** [comment]
Sync isn’t supported in Firefox for Android. Using `(browser.storage.sync || browser.storage.local)` lets it use local storage instead.

**Comment by da2x** [comment]
After a quick review: Sync support should be as easy as replacing `storage.local` with `storage.sync` in two locations **plus** migrating users from local to sync storage by moving all entries from one to the other.

**Comment by doronbehar** [comment]
Yea, I think what holds back @kintesh from just replacing `storage.local` with `storage.sync` is the migration experience - he doesn't want users to have an automatic update to the extension (as it is happens by default just in the background) and find out that all of their rules just disappeared.

**Comment by pckilgore** [comment]
@kintesh I'd like to help with this as its one of my biggest headaches.  I'll create a PR for further discussion, prefer I make it against dev, or master?

**Comment by subsign** [comment]
> Yea, I think what holds back @kintesh from just replacing `storage.local` with `storage.sync` is the migration experience - he doesn't want users to have an automatic update to the extension (as it is happens by default just in the background) and find out that all of their rules just disappeared.

Maybe use both `storage.local` and `storage.sync` and have the user choose? Or the much cooler option of having a diff-style editor to merge from local to sync. The editor offers two core benefits; one allows for the user to manage local to sync and deal with any merge conflicts and the other benefit is an editor for container mappings.

I got the idea from uBlock Origins:
[uBlock/dyna-rules.html at master · gorhill/uBlock · GitHub](https://github.com/gorhill/uBlock/blob/master/src/dyna-rules.html)
[uBlock/dyna-rules.js at master · gorhill/uBlock · GitHub](https://github.com/gorhill/uBlock/blob/master/src/js/dyna-rules.js)

![46023462-b2b94c80-c0b2-11e8-87c0-9fa821563364.png](https://user-images.githubusercontent.com/585534/46023462-b2b94c80-c0b2-11e8-87c0-9fa821563364.png)

--- 
Now in addition to CSV, rules can be added for mappings and even better, enforced. 

[CodeMirror: User Manual - Writing CodeMirror Modes](https://codemirror.net/doc/manual.html#modeapi) 
>Modes typically consist of a single JavaScript file. This file defines, in the simplest case, a lexer (tokenizer) for your language—a function that takes a character stream as input, advances it past a token, and returns a style for that token. More advanced modes can also handle indentation for the language.

As can be seen here for example:
[uBlock/ubo-dynamic-filtering.js at master · gorhill/uBlock · GitHub](https://github.com/gorhill/uBlock/blob/master/src/js/codemirror/ubo-dynamic-filtering.js)
```
CodeMirror.defineMode('ubo-dynamic-filtering', ( ) => {

    const validSwitches = new Set([
        'no-strict-blocking:',
        'no-popups:',
        'no-cosmetic-filtering:',
        'no-remote-fonts:',
        'no-large-media:',
        'no-csp-reports:',
        'no-scripting:',
    ]);
    const validSwitcheStates = new Set([
        'true',
        'false',
    ]);
```
Wiki page for rule syntax:
[Dynamic filtering: rule syntax · gorhill/uBlock Wiki · GitHub](https://github.com/gorhill/uBlock/wiki/Dynamic-filtering:-rule-syntax)

[CodeMirror: merge view demo](https://codemirror.net/demo/merge.html)
>The merge addon provides an interface for displaying and merging diffs, either two-way or three-way. The left (or center) pane is editable, and the differences with the other pane(s) are optionally shown live as you edit it. In the two-way configuration, there are also options to pad changed sections to align them, and to collapse unchanged stretches of text.

**Comment by rcorrear** [comment]
Was there ever a consensus about this? I started using this extension recently and I'd like to have this feature as I move between computers regularly (I could contribute code as well if there's a consensus on what to do).

**Comment by LoveIsGrief** [comment]
@rcorrear 
I don't think so, but I'd say if you find a solution that works for sync users and non-sync users (like me), just create a pull-request and @kintesh might have a look at it. I'll surely have a look at it too, since the implementation would be interesting.

It would most likely be necessary to modify the Storage wrapper to use the correct instance depending on a user setting.

**Comment by k4n30** [comment]
Worth taking into consideration - https://blog.mozilla.org/addons/2020/07/09/changes-to-storage-sync-in-firefox-79/

**Comment by rcorrear** [comment]
Just FYI, I decided not to work on this as Conex actually fits my needs better with the "ask in which container to open external links" option enabled.

**Comment by OmenWild** [comment]
I am still sticking with Containerise, and would love to have sync'd storage.

**Comment by robertl87** [comment]
The possibility to export a plain text list is indeed a very nice feature to export it to another computer, but I'd like to do it automatically (by ff sync or having the possibility to backup a file in my PC's appdata dir or something).

**Comment by rtevans01** [comment]
Is there an update on this feature request? Did anyone make a pull request yet?

**Comment by pixiekat** [comment]
If the sync option is too difficult, could the extension have an import/export to json option (like Sidebery, uBlock, etc)? At least then we can easily transfer rules to other devices.
